backend-app/server/controllers/authController.js

```diff
@@ -1,50 +1,133 @@
-const { check, validationResult } = require('express-validator');
 const User = require('../models/user');
+const { check, validationResult } = require('express-validator');
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcryptjs');
 
-// authMiddlewarefor registration
-const validateUser = [
-    check('userName', 'Username is required').not().isEmpty(),
-    check('email', 'Email is required').not().isEmpty(),
-    check('password', 'Password must be at least 6 characters long').isLength({ min: 6 }),
+// authMiddleware for registration
+const validateSignUp = [
+    check('username').trim().isEmpty().withMessage('Username is required')
+    .isLength({ min: 3 }).withMessage('Username must be at least 3 characters long'),
+    check('email').trim().isEmail().withMessage( 'Valid email is required'),
+    check('password').isLength({ min: 6}).withMessage('Password must be at least 6 characters long'),
 ];
 
-const registerUser = async (req, res) => {
-    // Check for validation errors
-    const errors = validationResult(req);
-    if (!errors.isEmpty()) {
-        return res.status(400).json({ errors: errors.array() });
-    }
+// Validation middleware for sign-in
+const validateSignIn = [
+    check('email').trim().isEmail().withMessage('Valid email is required'),
+    check('password').not().isEmpty().withMessage('Password is required'),
+];
+
+// User sign-up
+exports.signUp = async (req, res) => {
     try {
-        const { userName, email, password } = req.body;
 
-        // Check if user already exits
-        const existingUser = await User.findOne({ email });
-        if (existingUser) {
-            return res.status(400).json({ error: 'User already exists'});
+        // Check for validation errors
+        const errors = validationResult(req);
+        if (!errors.isEmpty()) {
+            return res.status(400).json({ errors: errors.array() });
         }
 
-        // Salting
-        const salt = await bcrypt.genSalt(10);
+        const { username, email, password, ...rest } = req.body;
 
-        // Hashing the password with salt
-        const hashedPassword = await bcrypt.hash(password, salt);
+        // Check for existing user with the same email
+        const existingEmailUser = await User.findOne({ email });
+        if (existingEmailUser) {
+            return res.status(400).json({ error: 'Email already in use' });
+        }
 
-        // Create a new user with the hashed password
+        // Check for existing user with the same username
+        const existingUsernameUser = await User.findOne({ username });
+        if (existingUsernameUser) {
+            return res.status(400).json({ error: 'Username already in use' });
+        }
+
+        // Salt and Hash password before saving
+        const salt = await bcrypt.genSalt(10);
+        const hashedPassword = await bcrypt.hash(password, salt);
         const user = new User({
-            userName,
             email,
```

Comment:
@Nhlanhla-advocate , please make sure this is , username, email, password *NOT* email,username, password